Preprocessing
===



In [1]:
from constants import *  # bad practice but YOLO
import preproc

In [2]:
import constants
reload(constants);
from constants import *

In [3]:
reload(preproc);
y_tr, x_tr = preproc.read_data(TRAIN_FILE, #filename
                              'subreddit', #label field
                              preprocessor=preproc.tokenize_and_downcase) #your preprocessor

y_te, x_te = preproc.read_data(TEST_FILE, #filename
                              'subreddit', #label field
                              preprocessor=preproc.tokenize_and_downcase) #your preprocessor

In [4]:
reload(preproc);
vocab = preproc.create_vocab(x_tr, x_te)

In [5]:
reload(preproc);
mapping = preproc.create_mapping(vocab)

In [6]:
# Now, we'll turn all of our input features into feature vectors
import numpy as np

Y_tr = np.array(y_tr)
X_tr = [preproc.feature_vector(bow, vocab, mapping) for bow in x_tr]

Y_te = np.array(y_te)
X_te = [preproc.feature_vector(bow, vocab, mapping) for bow in x_te]

In [23]:
# It looks like our training data may be too big and might cause memory issues
# we'll chop 1/2 of our training data off for now
X_tr_orig = X_tr
Y_tr_orig = Y_tr
# N = len(Y_tr)
# Y_tr = Y_tr_orig[:5*N/6]
# X_tr = X_tr_orig[:5*N/6]

Learning
=========

In [7]:
import learner

In [8]:
reload(learner);

nb = learner.trainNB(X_tr, Y_tr)

In [9]:
print learner.accuracy(nb, X_tr, Y_tr)
print learner.accuracy(nb, X_te, Y_te)

0.8575
0.698


Learning Results
=======


So far, we've got:
- 85.7% training accuracy
- 69.8% test accuracy

Can we do better? Let's try dimensionality reduction


In [45]:
N = len(Y_tr_orig)
X_tr = X_tr_orig[:N/20]
Y_tr = Y_tr_orig[:N/20]
print len(X_tr)
print len(X_te)

200
500


In [43]:
reload(learner);
X_tr = X_tr_orig
rp = learner.get_RP_features(X_tr, X_te, vocab)
#     pca_X_tr, pca_X_te

In [44]:
rp_X_tr = rp.transform(X_tr)
rp_X_te = rp.transform(X_te)

MemoryError: 

In [21]:
print X_tr[0]
print len(X_tr[0])
print ica_X_tr[0]
print len(ica_X_tr[0])

[ 0.  0.  0. ...,  0.  0.  0.]
29446
[ 0.05401611 -0.05252075  0.07767487  0.04510498 -0.05822754 -0.09094238
  0.02764893  0.04852295  0.06116819  0.05700684 -0.07849121  0.06561279
 -0.05753326 -0.07012939  0.04693604 -0.07090759  0.06347656 -0.05761719
  0.07592773  0.05029297  0.08459473 -0.08486938 -0.08197021  0.05651855
 -0.05078125 -0.03518677 -0.08355713 -0.0769043  -0.05456543  0.06204224
  0.05822754  0.04852295 -0.06497192 -0.07269287  0.06738281 -0.06280518
 -0.046875    0.08325195 -0.0473938   0.04724121 -0.06958008 -0.05718994
 -0.0447998  -0.06204224  0.03684235  0.03671265  0.05596924  0.06611633
 -0.06036377  0.05767822 -0.06237793 -0.05093384 -0.01528931 -0.06274414
 -0.07092285  0.06771851  0.06109619  0.06677246  0.07867432  0.05722046
 -0.07055664  0.03881836  0.06494141  0.07537842 -0.06838989 -0.07006073
  0.07330322 -0.06555176 -0.08850098 -0.07875824 -0.08163452  0.0949707
  0.04806519  0.04998779 -0.07006836  0.06402588 -0.06903076  0.06524658
  0.04937744  0